.github/workflows/container-image.yaml

```diff
@@ -1,53 +1,79 @@
-    - name: Set up Docker Buildx
-      id: buildx
-      uses: crazy-max/ghaction-docker-buildx@v3
-      with:
-        buildx-version: latest
-        qemu-version: latest
-    - name: Build container image
-      run: |
-        docker buildx build \
-          -t aws-ebs-csi-driver \
-          --platform=linux/arm64,linux/amd64 \
-          --output="type=image,push=false" . \
-          --target=amazonlinux
-    - name: Push to Github registry
-      run: |
-        USER=$(echo $GITHUB_REPOSITORY | cut -d'/' -f1)
-        BRANCH=$(echo $GITHUB_REF | cut -d'/' -f3)
-        IMAGE=aws-ebs-csi-driver
-        if [ "$BRANCH" = "master" ]; then
-          TAG="latest"
-        else
-          TAG=$BRANCH
-        fi
-        docker login docker.pkg.github.com -u $USER -p ${{ secrets.REGISTRY_TOKEN }}
-        docker build -t aws-ebs-csi-driver . --target amazonlinux
-        docker tag aws-ebs-csi-driver docker.pkg.github.com/$GITHUB_REPOSITORY/$IMAGE:$TAG
-        docker push docker.pkg.github.com/$GITHUB_REPOSITORY/$IMAGE:$TAG
-    - name: Push to Dockerhub registry
-      run: |
-        BRANCH=$(echo $GITHUB_REF | cut -d'/' -f3)
-        REPO=amazon/aws-ebs-csi-driver
-        if [ "$BRANCH" = "master" ]; then
-          TAG="latest"
-        else
-          TAG=$BRANCH
-        fi
-        docker login -u ${{ secrets.DOCKERHUB_USER }} -p ${{ secrets.DOCKERHUB_TOKEN }}
-        docker buildx build \
-          -t $REPO:$TAG \
-          --platform=linux/arm64,linux/amd64 \
-          --output="type=image,push=true" . \
-          --target=amazonlinux
-
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Build debian target
+        run: |
+          docker buildx build \
+            -t $DEB_BUILD_TAG \
+            --platform=linux/arm64,linux/amd64 \
+            --output="type=image,push=false" .
+            --target=debian-base
+      - name: Build amazonlinux target
+        run: |
+          docker buildx build \
+            -t $AL2_BUILD_TAG \
+            --platform=linux/arm64,linux/amd64 \
+            --output="type=image,push=false" . \
+            --target=amazonlinux
+      - name: Set environment variables
+        run: |
+          BRANCH=$(echo $GITHUB_REF | cut -d'/' -f3)
+          echo "BRANCH=$BRANCH" >> $GITHUB_ENV
+          if [ "$BRANCH" = "master" ]; then
+            TAG="latest"
+          else
+            TAG=$BRANCH
+          fi
+          echo "TAG=$TAG" >> $GITHUB_ENV
+      - name: Login to GitHub
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.REGISTRY_TOKEN }}
+      - name: Push debian target to GitHub
+        run: |
+          DEB_PUSH_TAG="docker.pkg.github.com/$GITHUB_REPOSITORY/$IMAGE:$TAG"
+          docker buildx build \
+            -t $DEB_PUSH_TAG \
```

Comment:
I don't know if this uses the cache from previous build step ($DEB_BUILD_TAG) . If not, there is no point to $DEB_BUILD_TAG/ $AL2_BUILD_TAG

Reply:
Yeah I'm a little bit confused. What's the benefit of build steps?

Reply:
makes it quicker to debug what failed. if build succeed but push step failed. etc.

Reply:
from local testing the build step image is cached so IMO it makes sense to separate the two steps